plugins/processors/snmp_lookup/lookup.go

```diff
@@ -0,0 +1,341 @@
+	l.cache = expirable.NewLRU[string, tagMap](l.CacheSize, nil, time.Duration(l.CacheTTL))
+	if l.Ordered {
+		l.parallel = parallel.NewOrdered(acc, l.addAsync, orderedQueueSize, l.ParallelLookups)
+	} else {
+		l.parallel = parallel.NewUnordered(acc, l.addAsync, l.ParallelLookups)
+	}
+	return nil
+}
+
+func (l *Lookup) Add(metric telegraf.Metric, _ telegraf.Accumulator) error {
+	l.parallel.Enqueue(metric)
+	return nil
+}
+
+func (l *Lookup) addAsync(metric telegraf.Metric) []telegraf.Metric {
+	if _, ok := metric.GetTag(l.AgentTag); !ok {
+		l.Log.Warn("Agent tag missing")
+		return []telegraf.Metric{metric}
+	}
+
+	index, ok := metric.GetTag(l.IndexTag)
+	if !ok {
+		l.Log.Warn("Index tag missing")
+		return []telegraf.Metric{metric}
+	}
+
+	gs, err := l.getConnection(metric)
+	if err != nil {
+		l.Log.Errorf("Could not prepare connection: %v", err)
+		return []telegraf.Metric{metric}
+	}
+
+	// Prepare cache
+	if err := l.prepareCache(gs, index); err != nil {
+		l.Log.Warnf("Could not prepare cache for %q: %v", gs.Host(), err)
+		return []telegraf.Metric{metric}
+	}
+
+	// Load from cache
+	tagMap, inCache := l.cache.Get(gs.Host())
+	tags, indexExists := tagMap.rows[index]
+	if inCache && indexExists {
+		for key, value := range tags {
+			metric.AddTag(key, value)
+		}
+	} else {
+		l.Log.Warnf("Could not find index %q on agent %q", index, gs.Host())
+	}
+
+	return []telegraf.Metric{metric}
+}
+
+// prepareCache prepares the cache if needed (index does not exist yet, or agent not yet cached)
+func (l *Lookup) prepareCache(gs snmpConnection, index string) error {
+	agent := gs.Host()
+
+	// Check cache
+	l.lock.Lock()
+	tagMap, inCache := l.cache.Peek(agent)
+	_, indexExists := tagMap.rows[index]
+
+	// Cache miss or non existing index and not recently refreshed the table
+	if !inCache || (!indexExists && time.Since(tagMap.created) > minRetry) {
+		// Check if another process is alreay loading the table, wait until done if so
+		if done, busy := l.sigs[agent]; busy {
+			l.lock.Unlock()
+			<-done
+		} else {
+			// No other process is already loading the table, let others know by creating a channel
+			l.sigs[agent] = make(chan struct{})
+			l.lock.Unlock()
+
+			if err := gs.Reconnect(); err != nil {
+				l.signalAgentReady(agent)
+				return fmt.Errorf("could not connect: %w", err)
+			}
+
+			// build the table for the configured tags
+			if err := l.loadTagMap(gs); err != nil {
+				l.signalAgentReady(agent)
+				return fmt.Errorf("could not load table: %w", err)
+			}
+
+			// Done loading, inform other processes
+			l.signalAgentReady(agent)
+		}
+	} else {
+		l.lock.Unlock()
+	}
+
+	return nil
+}
+
+func (l *Lookup) signalAgentReady(agent string) {
+	l.lock.Lock()
+	close(l.sigs[agent])
+	delete(l.sigs, agent)
+	l.lock.Unlock()
+}
+
+// getConnectionNoMock prepares a snmpConnection from the given metric tags (if present)
+func (l *Lookup) getConnectionNoMock(metric telegraf.Metric) (snmpConnection, error) {
```

Comment:
```suggestion
func (l *Lookup) getConnection(metric telegraf.Metric) (snmpConnection, error) {
```

Reply:
Doesn't that conflict with https://github.com/influxdata/telegraf/pull/14223/files#diff-47aec56cba3b4f326e33160527caaafed246fee74bdba44c0c792e0dc276f836R72?

Reply:
@srebhan ?

Reply:
Is this what you wanted? [01a2b5b](https://github.com/influxdata/telegraf/pull/14223/commits/01a2b5bd6ec46a782e2d9a5cc46690ea573e7b1d)